cmd/kube-apiserver/app/server.go

```diff
@@ -244,14 +245,16 @@ func CreateNodeDialer(s completedServerRunOptions) (tunneler.Tunneler, *http.Tra
 		// Get ssh key distribution func, if supported
 		var installSSHKey tunneler.InstallSSHKey
 
-		cloudprovider.DeprecationWarningForProvider(s.CloudProvider.CloudProvider)
-		cloud, err := cloudprovider.InitCloudProvider(s.CloudProvider.CloudProvider, s.CloudProvider.CloudConfigFile)
-		if err != nil {
-			return nil, nil, fmt.Errorf("cloud provider could not be initialized: %v", err)
-		}
-		if cloud != nil {
-			if instances, supported := cloud.Instances(); supported {
-				installSSHKey = instances.AddSSHKeyToAllInstances
+		if !utilfeature.DefaultFeatureGate.Enabled(features.DisableCloudProviders) {
```

Comment:
Wondering if we should be more aggressive and just exit early whenever --cloud-provider is set to either aws, azure, gce, openstack or vsphere. Would just require adding these checks across the code base:

```go
if utilfeature.DefaultFeatureGate.Enabled(features.DisableCloudProviders) && cloudprovider.IsNotExternal(s.CloudProvider.CloudProvider) {
    klog.Fatalf("...")
}
```

Personally, I don't think we want to disable this silently, better to fail early and let users know this functionality is not enabled.

Reply:
@andrewsykim Sounds good to me. I still like to stick to centralized implementation in `cloudprovider.IsExternal` method then throwing different error logs in each binary. It is better then to upgrade and realize that your cluster is no longer connected to cloud. Updated with a error message, do you think it looks better now? https://github.com/kubernetes/kubernetes/pull/100136/commits/6dd3d7b35d642f82da19da1e36ea31f63a4eaf92

Reply:
I'd rather do it at the component level, because cloud provider package is shared by external consumers who do not care about the in-tree behaviors.

Reply:
Seems like @deads2k agrees https://github.com/kubernetes/kubernetes/pull/100136#discussion_r625283514